### PR TITLE
Add configuration to lint missing docs of `pub(crate)` items

### PR DESCRIPTION
**Pull request by pvdrz**
Fixes this: https://github.com/rust-lang/rust-clippy/issues/5736#issuecomment-1412442404

TODO:
- [x] Needs docs
- [x] Needs better names
- [x] Should `pub` items be checked to when this new option is enabled? I'm saying no because `missing_docs` already exists

@flip1995 I'd like to get some input from you :) 

---

changelog: Enhancement: [`missing_docs_in_private_items`]: Added new configuration `missing-docs-in-crate-items` to lint on items visible within the current crate. For example, `pub(crate)` items.
[#10303](https://github.com/rust-lang/rust-clippy/pull/10303)

### CODE REVIEW

**Comment by flip1995** [comment]
> * Should `pub` items be checked to when this new option is enabled? I'm saying no because `missing_docs` already exists

Agreed. I also am fine with the name. I'll let @giraffate do the final review, I only skimmed over the name and documentation.

File: tests/ui-toml/pub_crate_missing_docs/pub_crate_missing_doc.rs
Comment (giraffate) [inline tests/ui-toml/pub_crate_missing_docs/pub_crate_missing_doc.rs:47]: question: This configuration says `Whether to **only** check for missing documentation in pub(crate) items`. Should the case below in top-level module be checked?
```suggestion
type Typedef = String;
pub type PubTypedef = String;

```

**Review by giraffate — COMMENTED** [review]
Thanks! I asked one question.

The `tests/ui-toml/pub_crate_missing_docs/pub_crate_missing_doc.stdout` file is empty, so it should be removed.

File: tests/ui-toml/pub_crate_missing_docs/pub_crate_missing_doc.rs
Comment (pvdrz) [reply tests/ui-toml/pub_crate_missing_docs/pub_crate_missing_doc.rs:47]: I have mixed feelings about this. On one hand `Typedef` is visible inside of the whole crate so it would be checked. However, some people might want to just check stuff that's marked explicitly as `pub(crate)` or `pub(super)` (where `super == crate`).

Comment (giraffate) [reply tests/ui-toml/pub_crate_missing_docs/pub_crate_missing_doc.rs:47]: IMO it's okay that `Typedef` is checked, but it seems to be confusing the note of configuration, for example how about the following:
`Whether to check for missing documentation in items visible within the current crate, for example pub(crate) items`

And can the tests like https://github.com/rust-lang/rust-clippy/pull/10303/files#r1106516655 be added?

Comment (pvdrz) [reply tests/ui-toml/pub_crate_missing_docs/pub_crate_missing_doc.rs:47]: sure I'll do both and ping you again

File: book/src/lint_configuration.md
Comment (giraffate) [inline book/src/lint_configuration.md:545]: Can this be updated in the same way as https://github.com/rust-lang/rust-clippy/pull/10303/files#diff-1d884fd439e15a9e47e1135cdc0ab5ee896a876b9f4c256ecce4be6c69a8831dR459?

File: clippy_lints/src/missing_doc.rs
Comment (giraffate) [inline clippy_lints/src/missing_doc.rs:39]: Can this be updated in the same way as https://github.com/rust-lang/rust-clippy/pull/10303/files#diff-1d884fd439e15a9e47e1135cdc0ab5ee896a876b9f4c256ecce4be6c69a8831dR459?

**Review by giraffate — APPROVED** [review]
Overall looks good, thanks!

I made small comments. I'll merge this once those are addressed.

**Comment by pvdrz** [comment]
Thanks @giraffate! It seems we're good to go then

**Comment by giraffate** [comment]
@bors r+

Thanks!

**Pull request merged**
